src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.Unix.cs

```diff
@@ -61,5 +61,97 @@ internal static unsafe void GetAnsiStringBytes(ReadOnlySpan<char> chars, Span<by
             int actualByteLength = Encoding.UTF8.GetBytes(chars, bytes);
             bytes[actualByteLength] = 0;
         }
+
+        public static IntPtr AllocHGlobal(IntPtr cb)
+        {
+            nuint cbNative = (nuint)(nint)cb;
+
+            // Avoid undefined malloc behavior by always allocating at least one byte
+            IntPtr pNewMem = Interop.Sys.MemAlloc((cbNative != 0) ? cbNative : 1);
+            if (pNewMem == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+
+            return pNewMem;
+        }
+
+        public static void FreeHGlobal(IntPtr hglobal)
+        {
+            if (hglobal != IntPtr.Zero)
```

Comment:
This check might be better in `MemFree`? It's fine to free a null pointer, but presumably we want to avoid the expensive managed->native transition, so putting this check closer to that covers more cases.

Reply:
Yep. Also, it matches what the code has done historically in CoreCLR.